reVX/offshore/dist_to_ports.py

```diff
@@ -0,0 +1,395 @@
+                                   np.radians(pixel_coords))
+        if len(port_coords) == 1:
+            dist = dist.flatten()
+        else:
+            dist = np.diag(dist)
+
+        # radius of the earth in kilometers # radius of the earth in km
+        R = 6371.0
+
+        return dist * R
+
+    @staticmethod
+    def _lc_dist_to_port(cost_arr, port_idx, port_dist):
+        """
+        Compute the least cost dist from the port coordinates to all
+        offshore coordinates in km
+
+        Parameters
+        ----------
+        cost_arr : ndarray
+            Cost array to compute least cost path to ports from. Array should
+            have values of 90m (pixel size) for offshore and 9999 for onshore
+        port_idx : list | tuple | ndarray
+            Port (row, col) index, used as starting point for least cost
+            distance
+        port_dist : float
+            Distance from port to pixel that corresponds to port_idx in meters
+
+        Returns
+        -------
+        lc_dist : ndarray
+            Least cost distance from port to all offshore pixels in km
+        """
+        logger.debug('Computing least cost distance from port that is {}km '
+                     'from pixel {} to all offshore pixels.'
+                     .format(port_dist, port_idx))
+        if not isinstance(port_idx, np.ndarray):
+            port_idx = np.array(port_idx)
+
+        if len(port_idx) == 2:
+            port_idx = np.expand_dims(port_idx, 0)
+
+        mcp = MCP_Geometric(cost_arr)
+        lc_dist, _ = mcp.find_costs(starts=port_idx)
+        lc_dist = lc_dist.astype('float32') / 1000
+        lc_dist += port_dist
+
+        return lc_dist
+
+    @classmethod
+    def _parse_ports(cls, ports, offshore_lat_lon):
+        """
+        Load ports and add mapping and distance to nearest offshore pixel
+
+        Parameters
+        ----------
+        ports : str
+            Path to shape file containing ports to compute least cost distance
+            to
+        offshore_lat_lon : pandas.DataFrame
+            Mapping of offshore pixel coordinates (lat, lon) to array indices
+            (row, col)
+
+        Returns
+        -------
+        ports : geopandas.GeoDataFrame
+            DataFrame of port locations and their mapping to the offshore
+            pixels for least cost distance computation
+        """
+        ports = gpd.read_file(ports, ignore_geometry=True)
+
+        pixel_coords = offshore_lat_lon[['latitude', 'longitude']].values
+        tree = cKDTree(pixel_coords)  # pylint: disable=not-callable
+
+        port_coords = ports[['LATITUDE', 'LONGITUDE']].values
+        _, idx = tree.query(port_coords)
+
+        pixels = offshore_lat_lon.iloc[idx]
+
+        ports['row'] = pixels['row'].values
+        ports['col'] = pixels['col'].values
+        ports['dist_to_pixel'] = cls._haversine_distance(port_coords,
+                                                         pixel_coords[idx])
```

Comment:
Is this mapping the ports to their respective nearest _single_ 90m pixels? If so, should specify in docstring. Current statement of "port locations and their mapping to the offshore pixels" makes it sound like it could be port -> 1 or more offshore pixels.

Reply:
not quite, it is to the nearest offshore 90m pixel. The issue is that sometimes the ports sit inland based on our dist_to_coast raster